File: slogger/src/main/java/io/github/heartinfei/slogger/Configuration.java
package io.github.heartinfei.slogger;

import android.content.Context;
import android.support.annotation.NonNull;

/**
 * 简介：配置信息
 *
 * @author 王强 on 2017/11/23 249346528@qq.com
 */
public class Configuration {
    private static final int DEFAULT_DEEP = Integer.MAX_VALUE;
    /**
     * 默认Tag
     */
    private String tag = "";
    /**
     * 堆栈深度
     */
    private int trackInfoDeep = DEFAULT_DEEP;

    /**
     * true 打印Tag
     */
    private boolean isPrintTag = true;

    /**
     * true 打印堆栈信息
     */
    private boolean isPrintTrackInfo = true;

    /**
     * true 打印行号
     */
    private boolean isPrintLineNo = true;

    /**
     * true 打印线程信息
     */
    private boolean showThreadInfo = true;

    /**
     *
     */
    private String targetContextName;

    /**
     * 应用程序包名
     */
    private String pkgName;

    public Configuration(Builder builder) {
        this.targetContextName = builder.targetContextName;
        this.tag = builder.tag;
        this.trackInfoDeep = builder.trackInfoDeep;
        this.isPrintTag = builder.isPrintTag;
        this.isPrintTrackInfo = builder.isPrintTrackInfo;
        this.isPrintLineNo = builder.isPrintLineNo;
        this.showThreadInfo = builder.showThreadInfo;
        this.pkgName = builder.pkgName;
    }

    public String getTargetContextName() {
        return targetContextName;
    }

    public String getPkgName() {
        return pkgName;
    }

    public String getTag() {
        return isPrintTag ? tag : "";
    }

    public boolean isShowThreadInfo() {
        return showThreadInfo;
    }

    public void setShowThreadInfo(boolean showThreadInfo) {
        this.showThreadInfo = showThreadInfo;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    public int getTrackInfoDeep() {
        return trackInfoDeep;
    }

    public void setTrackInfoDeep(int trackInfoDeep) {
        this.trackInfoDeep = trackInfoDeep;
    }

    public boolean isPrintTag() {
        return isPrintTag;
    }

    public void setPrintTag(boolean printTag) {
        isPrintTag = printTag;
    }

    public boolean isPrintLineNo() {
        return isPrintLineNo;
    }

    public void setPrintLineNo(boolean printLineNo) {
        isPrintLineNo = printLineNo;
    }

    public boolean isPrintTrackInfo() {
        return isPrintTrackInfo;
    }

    public void setPrintTrackInfo(boolean printTrackInfo) {
        isPrintTrackInfo = printTrackInfo;
    }

    public static final class Builder {
        private String tag;
        private int trackInfoDeep = DEFAULT_DEEP;
        private boolean isPrintTrackInfo = false;
        private boolean isPrintTag = true;
        private boolean isPrintLineNo = true;
        private boolean showThreadInfo = true;
        private String targetContextName;
        private final String pkgName;

        public Builder(@NonNull Context ctx) {
            this.targetContextName = ctx.getClass().getName();
            this.pkgName = ctx.getPackageName();
            this.tag = ctx.getClass().getSimpleName();
        }

        public Builder(@NonNull Context ctx, String tag) {
            this.targetContextName = ctx.getClass().getName();
            this.pkgName = ctx.getPackageName();
            this.tag = tag;
        }

        private boolean checkTag(String tag) {
            //不允许特殊字符
            return true;
        }

        public Builder tag(String val) {
            this.tag = val;
            return this;
        }

        public Builder isPrintTrackInfo(boolean val) {
            this.isPrintTrackInfo = val;
            return this;
        }

        public Builder trackInfoDeep(int deep) {
            trackInfoDeep = deep;
            return this;
        }

        public Builder isPrintTag(boolean val) {
            isPrintTag = val;
            return this;
        }

        public Builder isPrintLineNo(boolean val) {
            isPrintLineNo = val;
            return this;
        }

        public Builder isShowThreadInfo(boolean show) {
            this.showThreadInfo = show;
            return this;
        }

        public Configuration build() {
            return new Configuration(this);
        }
    }
}

File: slogger/src/main/java/io/github/heartinfei/slogger/Configuration.kt
package io.github.heartinfei.slogger

/**
 * [trackFilter] Used to filter stack
 * [tag]    Log withTag
 * [trackDeep] The stack deep to print.
 * [printTrackInfo] true print stack info.
 * [printThreadInfo] true print thread info
 * [printTimeStamp] true print timestamp.
 * @author Rango on 2019-05-29 249346528@qq.com
 */
class Configuration(
        var trackFilter: String? = null,
        var tag: String? = "",
        var trackDeep: Int = 1,
        var printTrackInfo: Boolean = true,
        var printThreadInfo: Boolean = true,
        var printTimeStamp: Boolean = false) {

    constructor(c: Configuration) : this(
            trackFilter = c.trackFilter,
            tag = c.tag,
            trackDeep = c.trackDeep,
            printTrackInfo = c.printTrackInfo,
            printThreadInfo = c.printThreadInfo,
            printTimeStamp = c.printTimeStamp
    )

    fun printThreadInfo(arg: Boolean): Configuration {
        this.printThreadInfo = arg
        return this
    }

    fun printTimeStamp(arg: Boolean): Configuration {
        this.printTimeStamp = arg
        return this
    }

    fun printTrackInfo(arg: Boolean): Configuration {
        this.printTrackInfo = arg
        return this
    }

    fun trackFilter(arg: String): Configuration {
        this.trackFilter = arg
        return this
    }

    fun trackDeep(deep: Int): Configuration {
        this.trackDeep = deep
        return this
    }

    fun tag(arg: String): Configuration {
        this.tag = arg
        return this
    }
}

/**
 * 链式调用封装
 */
class LogPrinterProxy(private val config: Configuration,
                      private val printer: LogPrinter) : LogPrinter by printer {

    fun setPrintTimeStamp(print: Boolean): LogPrinterProxy {
        config.printTimeStamp = print
        return this
    }

    fun setTrackFilter(trackFilter: String?): LogPrinterProxy {
        config.trackFilter = trackFilter
        return this
    }

    fun setTag(tag: String): LogPrinterProxy {
        config.tag = tag
        return this
    }

    fun setPrintTrackInfo(print: Boolean): LogPrinterProxy {
        config.printTrackInfo = print
        return this
    }

    fun setTrackDeep(level: Int): LogPrinterProxy {
        config.trackDeep = level
        return this
    }

    fun setPrintThreadInfo(print: Boolean): LogPrinterProxy {
        config.printThreadInfo = print
        return this
    }
}